src/Conf.cpp

```diff
@@ -358,6 +370,71 @@ void Conf::setDataCenter(const ConfParser::Node* node)
     }
 }
 
+void Conf::setCustomCommand(const ConfParser::Node* node)
+{
+    if (!node->sub) {
+        Throw(InvalidValue, "%s:%d CustomCommand require scope value", node->file, node->line);
+    }
+    for (auto p = node->sub; p; p = p->next) {
+        mCustomCommands.push_back(CustomCommandConf{});
+        auto& cc = mCustomCommands.back();
+        CustomCommandConf::init(cc, p->key.c_str(), Command::Sentinel);
+        auto s = p->sub;
+        if (!s) {
+            Throw(InvalidValue, "%s:%d CustomCommand.Command require scope value", node->file, node->line);
+        }
+        for (;s ; s = s->next) {
+            if (setInt(cc.cmd.minArgs, "minArgs", s, 2)) {
+            } else if (setInt(cc.cmd.maxArgs, "maxArgs", s, 2, 9999)) {
+            } else if (setCommandMode(cc.cmd.mode, "mode", s)) {
+            } else {
+                Throw(UnknownKey, "%s:%d unknown key %s", s->file, s->line, s->key.c_str());
+            }
+        }
+        Command::addCustomCommand(&cc.cmd);
+    }
+}
+
+bool Conf::setCommandMode(int& mode, const char* name, const ConfParser::Node* n, const int defaultMode)
+{
+    if (strcasecmp(name, n->key.c_str()) != 0) {
+        return false;
+    }
+
+    if (n->val.size() == 0) {
+        mode = defaultMode;
+    } else {
+        mode = 0;
+        int used = 0;
+        std::string mask;
+        std::istringstream is(n->val);
+        while (std::getline(is, mask, ',')) {
+            if ((strcasecmp(mask.c_str(), "Write") == 0)) {
+                used++;
+                mode |= Command::Write;
+            } else if ((strcasecmp(mask.c_str(), "Read") == 0)) {
+                used++;
+                mode |= Command::Read;
+            } else if ((strcasecmp(mask.c_str(), "Admin") == 0)) {
+                used++;
+                mode |= Command::Admin;
+            } else if ((strcasecmp(mask.c_str(), "KeyAt2") == 0)) {
+                mode |= Command::KeyAt2;
+            } else if ((strcasecmp(mask.c_str(), "KeyAt3") == 0)) {
+                mode |= Command::KeyAt3;
+            } else {
+                Throw(InvalidValue, "%s:%d unknown mode %s", n->file, n->line, mask.c_str());
+            }
+        }
+        if (used == 0) {
+            Throw(InvalidValue, "%s:%d %s command require exclusive mode", n->file, n->line, name);
+        } else if (used > 1) {
+            Throw(InvalidValue, "%s:%d %s command duplicated exclusive mode", n->file, n->line, name);
+        }
```

Comment:
not robust, consider below cases:
+ mode KeyAt2
+ mode Read,Read

By the way, I think use '|' to replace ',' as seperator is better

Reply:
I think
+ mode KeyAt2
mode should have default value Write, so it's final value is: Write|KeyAt2
+ mode Read|Read
isn't a good config, but we can let it pass.

maybe like below:
```
    switch (mode & Command::AuthMask) {
    case 0:
        mode |= Command::Write;
        break;
    case Command::Read:
    case Command::Write:
    case Command::Admin:
        break;
    default:
        Throw(...);
    }
```

Reply:
Thanks. All applied.